core/sile.lua

```diff
@@ -240,10 +241,251 @@ function SILE.resolveFile(filename, pathprefix)
   return nil
 end
 
+SILE.currentCommandStack = {
+  --- Internal: push-pop balance checking ID
+  _lastPushId = 0,
+  --- Internal: Function assigned to frames to convert them to string
+  _frameToString = function(self, skipFile, withAttrs)
+    local str
+    if skipFile or not self.file then
+      str = ""
+    else
+      str = self.file .. " "
+    end
+
+    if self.line then
+      str = str .. "l." .. self.line .. " "
+      if self.column then
+        str = str .. "col." .. self.column .. " "
+      end
+    end
+
+    if self.tag then
+      -- Command
+      str = str .. "\\" .. self.tag
+      if withAttrs then
+        str = str .. "["
+        local first = true
+        for key, value in pairs(self.options) do
+          if first then first = false else str = str .. ", " end
+          str = str .. key .. "=" .. value
+        end
+        str = str .. "]"
+      end
+    elseif self.text then
+      -- Literal string
+      local text = self.text
+      if text:len() > 20 then
+        text = text:sub(1, 18) .. "…"
+      end
+      text = text:gsub("\n", "␤"):gsub("\t", "␉"):gsub("\v", "␋")
+
+      str = str .. "\"" .. text .. "\""
+    else
+      -- Unknown
+      str = str .. type(self.content) .. "(" .. self.content .. ")"
+    end
+
+    --str = str .. "\n\n" .. self.pushTrace .. "\n"
+
+    return str
+  end
+}
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Carries information about the command call, not the command itself.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushCommand(tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or "???",
+    file = file or SILE.currentlyProcessingFile,
+    line = line,
+    column = column,
+    options = options or {}
+  })
+end
+--- Push a command frame on to the stack to record the execution trace for debugging.
+--- Command arguments are inferred from AST content, any item may be overridden.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushContent(content, tag, line, column, options, file)
+  return self:_pushFrame({
+    tag = tag or (type(content) == "table" and content.tag) or "???",
+    file = file or (type(content) == "table" and content.file) or SILE.currentlyProcessingFile,
+    line = line or (type(content) == "table" and content.line),
+    column = column or (type(content) == "table" and content.col),
+    options = options or (type(content) == "table" and content.attr) or {}
+  })
+end
+--- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+--- Must be popped with `pop(returnOfPush)`.
+function SILE.currentCommandStack:pushText(text, line, column, file)
```

Comment:
If we're pushing text on it, then this isn't exactly a command stack is it? Also if it's a stack then it is more than just "current". Maybe it should be renamed `SILE.contentStack` or something generic like that.

Reply:
If you consider literal text to be a command to typeset that text, then it can be a command. But then, if we allow to push any information onto the stack after merging with breadcrumb system, it won't even be a `contentStack`. Naming things is hard.

I would however still consider the `current` to be correct, as the items lower down on the stack are still being currently processed.

Then the name would be `currentStack`, which is still weird. `traceStack`?

Reply:
Much better naming...